# 🤖 Model Building

This notebook trains baseline and advanced regression models for taxi trip duration prediction.

Models:
- Linear Regression
- Decision Tree Regressor
- Random Forest Regressor
- XGBoost Regressor

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor

from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

import joblib
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv("../data/processed/train_featured.csv")

print("Dataset Shape:", df.shape)
df.head()

Dataset Shape: (1447928, 20)


,vendor_id,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration,pickup_hour,pickup_day,pickup_month,pickup_weekday,is_weekend,is_rush_hour,is_night,haversine_distance_km,manhattan_distance_km,longitude_difference,latitude_difference,bearing
0,2,1,-73.982155,40.767937,-73.964630,40.765602,0,455,17,14,3,0,0,1,0,1.498521,1.735433,0.017525,-0.002335,99.970196
1,1,1,-73.980415,40.738564,-73.999481,40.731152,0,663,0,12,6,6,1,0,1,1.805507,2.430506,-0.019066,-0.007412,242.846232
2,2,1,-73.979027,40.763939,-74.005333,40.710087,0,2124,11,19,1,1,0,0,0,6.385098,8.203575,-0.026306,-0.053852,200.319835
3,2,1,-74.010040,40.719971,-74.012268,40.706718,0,429,19,6,4,2,0,1,0,1.485498,1.661331,-0.002228,-0.013252,187.262300
4,2,1,-73.973053,40.793209,-73.972923,40.782520,0,435,13,26,3,5,1,0,0,1.188588,1.199457,0.000130,-0.010689,179.473585


In [4]:
X = df.drop("trip_duration", axis=1)
y = df["trip_duration"]

print("Features Shape:", X.shape)
print("Target Shape:", y.shape)

Features Shape: (1447928, 19)
Target Shape: (1447928,)


In [5]:
# Step 1: Separate 15% for final testing
X_temp, X_test, y_temp, y_test = train_test_split(
    X,
    y,
    test_size=0.15,
    random_state=42
)

# Step 2: Split remaining data into train and validation
X_train, X_valid, y_train, y_valid = train_test_split(
    X_temp,
    y_temp,
    test_size=0.15 / 0.85,
    random_state=42
)

print("X_train:", X_train.shape)
print("X_valid:", X_valid.shape)
print("X_test:", X_test.shape)

X_train: (1013548, 19)
X_valid: (217190, 19)
X_test: (217190, 19)


In [6]:
def evaluate_model(model, X_valid, y_valid):
    predictions = model.predict(X_valid)
    
    mae = mean_absolute_error(y_valid, predictions)
    rmse = np.sqrt(mean_squared_error(y_valid, predictions))
    r2 = r2_score(y_valid, predictions)
    
    return mae, rmse, r2

In [11]:
linear_model = LinearRegression()

linear_model.fit(X_train, y_train)

linear_mae, linear_rmse, linear_r2 = evaluate_model(
    linear_model,
    X_valid,
    y_valid
)

print("Linear Regression Results")
print("MAE:", linear_mae)
print("RMSE:", linear_rmse)
print("R2 Score:", linear_r2)

Linear Regression Results
MAE: 274.3553514713531
RMSE: 413.59255942231124
R2 Score: 0.611595329227681


In [7]:
decision_tree_model = DecisionTreeRegressor(
    random_state=42,
    max_depth=15
)

decision_tree_model.fit(X_train, y_train)

dt_mae, dt_rmse, dt_r2 = evaluate_model(
    decision_tree_model,
    X_valid,
    y_valid
)

print("Decision Tree Results")
print("MAE:", dt_mae)
print("RMSE:", dt_rmse)
print("R2 Score:", dt_r2)

Decision Tree Results
MAE: 209.02425749438058
RMSE: 354.7968477183837
R2 Score: 0.7141761171282495


In [13]:
random_forest_model = RandomForestRegressor(
    n_estimators=100,
    max_depth=20,
    random_state=42,
    n_jobs=-1
)

random_forest_model.fit(X_train, y_train)

rf_mae, rf_rmse, rf_r2 = evaluate_model(
    random_forest_model,
    X_valid,
    y_valid
)

print("Random Forest Results")
print("MAE:", rf_mae)
print("RMSE:", rf_rmse)
print("R2 Score:", rf_r2)

Random Forest Results
MAE: 186.61154573798808
RMSE: 309.0810993518537
R2 Score: 0.7830878364432142


In [14]:
xgb_model = XGBRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=8,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    objective="reg:squarederror",
    n_jobs=-1
)

xgb_model.fit(X_train, y_train)

xgb_mae, xgb_rmse, xgb_r2 = evaluate_model(
    xgb_model,
    X_valid,
    y_valid
)

print("XGBoost Results")
print("MAE:", xgb_mae)
print("RMSE:", xgb_rmse)
print("R2 Score:", xgb_r2)

XGBoost Results
MAE: 178.50921630859375
RMSE: 298.17255909959255
R2 Score: 0.7981287837028503


In [15]:
model_results = pd.DataFrame({
    "Model": [
        "Linear Regression",
        "Decision Tree",
        "Random Forest",
        "XGBoost"
    ],
    "MAE": [
        linear_mae,
        dt_mae,
        rf_mae,
        xgb_mae
    ],
    "RMSE": [
        linear_rmse,
        dt_rmse,
        rf_rmse,
        xgb_rmse
    ],
    "R2 Score": [
        linear_r2,
        dt_r2,
        rf_r2,
        xgb_r2
    ]
})

model_results.sort_values(by="RMSE")

,Model,MAE,RMSE,R2 Score
3,XGBoost,178.509216,298.172559,0.798129
2,Random Forest,186.611546,309.081099,0.783088
1,Decision Tree,209.024257,354.796848,0.714176
0,Linear Regression,274.355351,413.592559,0.611595


In [16]:
import os
import joblib

# Create models directory if it doesn't exist
os.makedirs("../models", exist_ok=True)

# Save all trained models
joblib.dump(linear_model, "../models/linear_regression.pkl")
joblib.dump(decision_tree_model, "../models/decision_tree.pkl")
joblib.dump(random_forest_model, "../models/random_forest.pkl")
joblib.dump(xgb_model, "../models/xgboost.pkl")

# Select best model based on lowest validation RMSE
best_model_name = model_results.sort_values(
    by="RMSE", ascending=True
).iloc[0]["Model"]

# Map model names to trained model objects
trained_models = {
    "Linear Regression": linear_model,
    "Decision Tree": decision_tree_model,
    "Random Forest": random_forest_model,
    "XGBoost": xgb_model
}

# Get best model
best_model = trained_models[best_model_name]

# Save best model
joblib.dump(best_model, "../models/best_model.pkl")

print(f"Best model selected: {best_model_name}")
print("All models and best model saved successfully!")

Best model selected: XGBoost
All models and best model saved successfully!


In [17]:
# Final evaluation on unseen test data

test_mae, test_rmse, test_r2 = evaluate_model(
    best_model,
    X_test,
    y_test
)

print("Final Test Results")
print("----------------------")
print(f"Best Model: {best_model_name}")
print(f"Test MAE: {test_mae:.2f} seconds")
print(f"Test RMSE: {test_rmse:.2f} seconds")
print(f"Test R2 Score: {test_r2:.4f}")

Final Test Results
----------------------
Best Model: XGBoost
Test MAE: 177.85 seconds
Test RMSE: 305.94 seconds
Test R2 Score: 0.7888


## Final Model Evaluation

Four regression algorithms were compared:
- Linear Regression
- Decision Tree
- Random Forest
- XGBoost

XGBoost achieved the lowest validation RMSE and was selected
as the final model.

### Performance on Unseen Test Data

- MAE: 177.85 seconds
- RMSE: 305.94 seconds
- R² Score: 0.7888

The final model was evaluated on a held-out test set
that was not used during model selection.

Further analysis is needed to identify large prediction
errors and assess generalization.

# Conclusion

Model building is complete.

Four regression models were trained and compared:

- Linear Regression
- Decision Tree
- Random Forest
- XGBoost

All trained models were saved in the `models/` directory.

The best-performing model was selected based on the lowest RMSE and saved as:

`models/best_model.pkl`

Next step: Model Evaluation and Hyperparameter Tuning.

In [19]:
print("Features used for training:")
for i, column in enumerate(X.columns, start=1):
    print(f"{i}. {column}")

Features used for training:
1. vendor_id
2. passenger_count
3. pickup_longitude
4. pickup_latitude
5. dropoff_longitude
6. dropoff_latitude
7. store_and_fwd_flag
8. pickup_hour
9. pickup_day
10. pickup_month
11. pickup_weekday
12. is_weekend
13. is_rush_hour
14. is_night
15. haversine_distance_km
16. manhattan_distance_km
17. longitude_difference
18. latitude_difference
19. bearing
